ibis/expr/rules.py

```diff
@@ -794,3 +799,127 @@ def _validate(self, args, i):
 
 def comparable(left, right):
     return ir.castable(left, right) or ir.castable(right, left)
+
+
+class Table(Argument):
+    """A table argument.
+
+    Parameters
+    ----------
+    name : str
+        The name of the table argument.
+    optional : bool
+        Whether this table argument is optional or not.
+    satisfying : iterable
+        An iterable of lambda expressions, which will be used to validate
+        arguments. Each lambda expression must take a table as its single
+        argument, validate, and then return ``True`` for tables that
+        pass validation and ``False`` otherwise.
+    schema : iterable
+        An iterable of column rules used to validate a table. Each member
+        must be an instance of ``Argument`` capable of validating a column
+        (most likely an object created with :func:`ibis.rules.column`). A
+        ``name`` attribute is required for each member, which will be used
+        to find a candidate column to satisfy the rule. A table must have
+        a columns to match each non-optional rule to pass validation.
+    allow_extra : bool
+        Only useful in conjunction with ``schema``. If set to ``False`` then
+        only tables with exactly the columns specified in the schema will be
+        allowed. If set to ``True`` then the table may have columns not
+        specified in the schema, as long as each non-optional member of the
+        schema was satisfied.
+    validator : ???
+        ???
+    doc : ???
+        ???
```

Comment:
@cpcloud I couldn't figure what to do with these (the `validator` and `doc` args). I don't use them but stuff breaks if I don't set them to `None`. Maybe I could set them to `None` manually (without turning them into args), but first I want to make sure that it makes sense. Do you know how these are used?

Reply:
`doc` lets you give a docstring to the argument:

```python
In [1]: class Foo(ibis.expr.operations.ValueOp):
   ...:     input_type = [ibis.expr.rules.double(name='foo', doc='bar bar double')]
   ...: 

In [2]: x = Foo(1)

In [3]: x.a?
Object `x.a` not found.

In [4]: x.foo?
Type:        property
String form: <property object at 0x7fedee9b2278>
Docstring:   bar bar double
```

Reply:
Looks like validators let you pass in your own validation function, see line 239 in this file.

Reply:
I see. Should we leave it? Can't think of any use case for it, if users want a custom validator they should just write one, right?

Reply:
bump @cpcloud